plan/gen_physical_plans.go

```diff
@@ -271,32 +283,122 @@ func (p *LogicalJoin) getIndexJoinByOuterIdx(prop *requiredProp, outerIdx int) [
 		pkCol := x.getPKIsHandleCol()
 		if pkCol != nil && innerJoinKeys[0].Equal(pkCol, nil) {
 			innerPlan := x.forceToTableScan()
-			return p.constructIndexJoin(prop, innerJoinKeys, outerJoinKeys, outerIdx, innerPlan)
+			return p.constructIndexJoin(prop, innerJoinKeys, outerJoinKeys, outerIdx, innerPlan, nil, nil)
 		}
 	}
+	var (
+		bestIndexInfo  *model.IndexInfo
+		rangesOfBest   []*ranger.IndexRange
+		maxUsedCols    int
+		remainedOfBest []expression.Expression
+	)
+	orderedInnerKeys := make([]*expression.Column, len(innerJoinKeys))
+	orderedOuterKeys := make([]*expression.Column, len(innerJoinKeys))
+	indexColOffsets := make([]int, 0, len(innerJoinKeys))
 	for _, indexInfo := range indices {
-		matchedOffsets := joinKeysMatchIndex(innerJoinKeys, indexInfo)
-		if matchedOffsets == nil {
+		isOk, remained, ranges, idxOff2KeyOff := p.checkIndexForIndexJoin(indexInfo, x, innerJoinKeys)
+		if !isOk {
 			continue
 		}
-		usedIndexInfo = indexInfo
-		newOuterJoinKeys := make([]*expression.Column, len(outerJoinKeys))
-		newInnerJoinKeys := make([]*expression.Column, len(innerJoinKeys))
-		for i, offset := range matchedOffsets {
-			newOuterJoinKeys[i] = outerJoinKeys[offset]
-			newInnerJoinKeys[i] = innerJoinKeys[offset]
+		// We choose the index by the number of used columns of the range, the much the better.
+		// Notice that there may be the cases like `t1.a=t2.a and b > 2 and b < 1`. So ranges can be nil though the conditions are valid.
+		// And obviously when the range is nil, we don't need index join.
+		if len(ranges) > 0 && len(ranges[0].LowVal) > maxUsedCols {
+			bestIndexInfo = indexInfo
+			maxUsedCols = len(ranges[0].LowVal)
+			rangesOfBest = ranges
+			remainedOfBest = remained
+
+			indexColOffsets = indexColOffsets[:0]
+			for key, val := range idxOff2KeyOff {
+				if val == -1 {
+					continue
+				}
+				indexColOffsets = append(indexColOffsets, key)
+			}
+			for i, idxOff := range indexColOffsets {
+				keyOff := idxOff2KeyOff[idxOff]
+				orderedOuterKeys[i] = outerJoinKeys[keyOff]
+				orderedInnerKeys[i] = innerJoinKeys[keyOff]
+			}
 		}
-		outerJoinKeys = newOuterJoinKeys
-		innerJoinKeys = newInnerJoinKeys
-		break
 	}
-	if usedIndexInfo != nil {
-		innerPlan := x.forceToIndexScan(usedIndexInfo)
-		return p.constructIndexJoin(prop, innerJoinKeys, outerJoinKeys, outerIdx, innerPlan)
+	if bestIndexInfo != nil {
+		innerPlan := x.forceToIndexScan(bestIndexInfo, remainedOfBest)
+		return p.constructIndexJoin(prop, orderedInnerKeys, orderedOuterKeys, outerIdx, innerPlan, rangesOfBest, indexColOffsets)
 	}
 	return nil
 }
 
+// checkIndexForIndexJoin checks whether this index can be used for building index join and return enough information to
+// determine whether this index is better and how to construct the index join.
+func (p *LogicalJoin) checkIndexForIndexJoin(indexInfo *model.IndexInfo, innerPlan *DataSource, innerJoinKeys []*expression.Column) (ok bool,
```

Comment:
It does more than check, 
How about name it `buildRangeForIndexJoin` and remove the bool return value?